### PR TITLE
fix CI: update to pybind 2.6.1

### PR DESCRIPTION
**Pull request by tigercosmos**

### CODE REVIEW

**Comment by tigercosmos** [comment]
怪惹 我mac可以動的說

File: CMakeLists.txt [hunk 1]
```diff
@@ -30,7 +30,7 @@ endif()
 
 include(Flake8)
 
-set(CMAKE_CXX_STANDARD 14)
+set(CMAKE_CXX_STANDARD 17)
```
Comment (yungyuc) [inline CMakeLists.txt]: It is intentional to use C++14.  Please revert this change,.

**Review by yungyuc — CHANGES_REQUESTED** [review]

**Comment by yungyuc** [comment]
There's still something wrong in CI.  I reran it but the failure is still there.

**Comment by tigercosmos** [comment]
I don't know why. It works fine on my mac. For reference:

```log
$ python3 setup.py build_ext           --cmake-args="-DCMAKE_BUILD_TYPE=Release -DPYTHON_EXECUTABLE=$(which python3)"           --make-args="VERBOSE=1"
running build_ext
-- The C compiler identification is AppleClang 12.0.0.12000032
-- The CXX compiler identification is AppleClang 12.0.0.12000032
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- use clang-tidy with DO_CLANG_TIDY: /usr/local/bin/clang-tidy
-- using program '/usr/local/bin/flake8'
-- HIDE_SYMBOL: OFF
-- DEBUG_SYMBOL: ON
-- MODMESH_PROFILE: OFF
-- Found PythonInterp: /usr/local/bin/python3 (found version "3.8.5") 
-- Found PythonLibs: /usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/libpython3.8.dylib
-- Performing Test HAS_FLTO
-- Performing Test HAS_FLTO - Success
-- Performing Test HAS_FLTO_THIN
-- Performing Test HAS_FLTO_THIN - Success
-- Found pybind11: /usr/local/include (found version "2.6.1" )
-- pybind11_INCLUDE_DIRS: /usr/local/include;/usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/include/python3.8
-- Found NumPy: /usr/local/lib/python3.8/site-packages/numpy/core/include (found version "1.19.4") 
-- NUMPY_INCLUDE_DIR: /usr/local/lib/python3.8/site-packages/numpy/core/include
-- CMAKE_INSTALL_INCLUDEDIR: include
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/tigercosmos/modmesh/build/temp.macosx-10.15-x86_64-3.8
/usr/local/Cellar/cmake/3.18.4/bin/cmake -S/Users/tigercosmos/modmesh -B/Users/tigercosmos/modmesh/build/temp.macosx-10.15-x86_64-3.8 --check-build-system CMakeFiles/Makefile.cmake 0
/Library/Developer/CommandLineTools/usr/bin/make  -f CMakeFiles/Makefile2 _modmesh
/usr/local/Cellar/cmake/3.18.4/bin/cmake -S/Users/tigercosmos/modmesh -B/Users/tigercosmos/modmesh/build/temp.macosx-10.15-x86_64-3.8 --check-build-system CMakeFiles/Makefile.cmake 0
/usr/local/Cellar/cmake/3.18.4/bin/cmake -E cmake_progress_start /Users/tigercosmos/modmesh/build/temp.macosx-10.15-x86_64-3.8/CMakeFiles 2
/Library/Developer/CommandLineTools/usr/bin/make  -f CMakeFiles/Makefile2 CMakeFiles/_modmesh.dir/all
/Library/Developer/CommandLineTools/usr/bin/make  -f CMakeFiles/_modmesh.dir/build.make CMakeFiles/_modmesh.dir/depend
cd /Users/tigercosmos/modmesh/build/temp.macosx-10.15-x86_64-3.8 && /usr/local/Cellar/cmake/3.18.4/bin/cmake -E cmake_depends "Unix Makefiles" /Users/tigercosmos/modmesh /Users/tigercosmos/modmesh /Users/tigercosmos/modmesh/build/temp.macosx-10.15-x86_64-3.8 /Users/tigercosmos/modmesh/build/temp.macosx-10.15-x86_64-3.8 /Users/tigercosmos/modmesh/build/temp.macosx-10.15-x86_64-3.8/CMakeFiles/_modmesh.dir/DependInfo.cmake --color=
Dependee "/Users/tigercosmos/modmesh/build/temp.macosx-10.15-x86_64-3.8/CMakeFiles/_modmesh.dir/DependInfo.cmake" is newer than depender "/Users/tigercosmos/modmesh/build/temp.macosx-10.15-x86_64-3.8/CMakeFiles/_modmesh.dir/depend.internal".
Dependee "/Users/tigercosmos/modmesh/build/temp.macosx-10.15-x86_64-3.8/CMakeFiles/CMakeDirectoryInformation.cmake" is newer than depender "/Users/tigercosmos/modmesh/build/temp.macosx-10.15-x86_64-3.8/CMakeFiles/_modmesh.dir/depend.internal".
Scanning dependencies of target _modmesh
/Library/Developer/CommandLineTools/usr/bin/make  -f CMakeFiles/_modmesh.dir/build.make CMakeFiles/_modmesh.dir/build
[ 50%] Building CXX object CMakeFiles/_modmesh.dir/src/modmesh.cpp.o
/Library/Developer/CommandLineTools/usr/bin/c++ -D_modmesh_EXPORTS -I/usr/local/lib/python3.8/site-packages/numpy/core/include -I/Users/tigercosmos/modmesh/include -isystem /usr/local/include -isystem /usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/include/python3.8 -g -O3 -DNDEBUG -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk -fPIC -fvisibility=default -Werror -Wall -Wextra -Wno-unused-value -Wno-noexcept-type -flto -std=c++14 -o CMakeFiles/_modmesh.dir/src/modmesh.cpp.o -c /Users/tigercosmos/modmesh/src/modmesh.cpp
[100%] Linking CXX shared module ../lib.macosx-10.15-x86_64-3.8/modmesh/_modmesh.cpython-38-darwin.so
/usr/local/Cellar/cmake/3.18.4/bin/cmake -E cmake_link_script CMakeFiles/_modmesh.dir/link.txt --verbose=1
/Library/Developer/CommandLineTools/usr/bin/c++  -g -O3 -DNDEBUG -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk -bundle -Wl,-headerpad_max_install_names -Xlinker -undefined -Xlinker dynamic_lookup -flto -o ../lib.macosx-10.15-x86_64-3.8/modmesh/_modmesh.cpython-38-darwin.so CMakeFiles/_modmesh.dir/src/modmesh.cpp.o 
/Library/Developer/CommandLineTools/usr/bin/strip -x /Users/tigercosmos/modmesh/build/lib.macosx-10.15-x86_64-3.8/modmesh/_modmesh.cpython-38-darwin.so
[100%] Built target _modmesh
/usr/local/Cellar/cmake/3.18.4/bin/cmake -E cmake_progress_start /Users/tigercosmos/modmesh/build/temp.macosx-10.15-x86_64-3.8/CMakeFiles 0
```

**Comment by tigercosmos** [comment]
it's fine even with lint:

```log
        python3 setup.py build_ext           --cmake-args="-DCMAKE_BUILD_TYPE=Release -DPYTHON_EXECUTABLE=$(which python3)"           --make-args="VERBOSE=1"
running build_ext
-- use clang-tidy with DO_CLANG_TIDY: ;-warnings-as-errors=*
-- using program '/usr/local/bin/flake8'
-- HIDE_SYMBOL: OFF
-- DEBUG_SYMBOL: ON
-- MODMESH_PROFILE: OFF
-- Found pybind11: /usr/local/include (found version "2.6.1" )
-- pybind11_INCLUDE_DIRS: /usr/local/include;/usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/include/python3.8
-- NUMPY_INCLUDE_DIR: /usr/local/lib/python3.8/site-packages/numpy/core/include
-- CMAKE_INSTALL_INCLUDEDIR: include
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/tigercosmos/modmesh/build/temp.macosx-10.15-x86_64-3.8
/usr/local/Cellar/cmake/3.18.4/bin/cmake -S/Users/tigercosmos/modmesh -B/Users/tigercosmos/modmesh/build/temp.macosx-10.15-x86_64-3.8 --check-build-system CMakeFiles/Makefile.cmake 0
/Library/Developer/CommandLineTools/usr/bin/make  -f CMakeFiles/Makefile2 _modmesh
/usr/local/Cellar/cmake/3.18.4/bin/cmake -S/Users/tigercosmos/modmesh -B/Users/tigercosmos/modmesh/build/temp.macosx-10.15-x86_64-3.8 --check-build-system CMakeFiles/Makefile.cmake 0
/usr/local/Cellar/cmake/3.18.4/bin/cmake -E cmake_progress_start /Users/tigercosmos/modmesh/build/temp.macosx-10.15-x86_64-3.8/CMakeFiles 2
/Library/Developer/CommandLineTools/usr/bin/make  -f CMakeFiles/Makefile2 CMakeFiles/_modmesh.dir/all
/Library/Developer/CommandLineTools/usr/bin/make  -f CMakeFiles/_modmesh.dir/build.make CMakeFiles/_modmesh.dir/depend
cd /Users/tigercosmos/modmesh/build/temp.macosx-10.15-x86_64-3.8 && /usr/local/Cellar/cmake/3.18.4/bin/cmake -E cmake_depends "Unix Makefiles" /Users/tigercosmos/modmesh /Users/tigercosmos/modmesh /Users/tigercosmos/modmesh/build/temp.macosx-10.15-x86_64-3.8 /Users/tigercosmos/modmesh/build/temp.macosx-10.15-x86_64-3.8 /Users/tigercosmos/modmesh/build/temp.macosx-10.15-x86_64-3.8/CMakeFiles/_modmesh.dir/DependInfo.cmake --color=
/Library/Developer/CommandLineTools/usr/bin/make  -f CMakeFiles/_modmesh.dir/build.make CMakeFiles/_modmesh.dir/build
[ 50%] Building CXX object CMakeFiles/_modmesh.dir/src/modmesh.cpp.o
/Library/Developer/CommandLineTools/usr/bin/c++ -D_modmesh_EXPORTS -I/usr/local/lib/python3.8/site-packages/numpy/core/include -I/Users/tigercosmos/modmesh/include -isystem /usr/local/include -isystem /usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/include/python3.8 -g -O3 -DNDEBUG -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk -fPIC -fvisibility=default -Werror -Wall -Wextra -Wno-unused-value -Wno-noexcept-type -std=c++14 -o CMakeFiles/_modmesh.dir/src/modmesh.cpp.o -c /Users/tigercosmos/modmesh/src/modmesh.cpp
[100%] Linking CXX shared module ../lib.macosx-10.15-x86_64-3.8/modmesh/_modmesh.cpython-38-darwin.so
/usr/local/Cellar/cmake/3.18.4/bin/cmake -E cmake_link_script CMakeFiles/_modmesh.dir/link.txt --verbose=1
/Library/Developer/CommandLineTools/usr/bin/c++  -g -O3 -DNDEBUG -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk -bundle -Wl,-headerpad_max_install_names -Xlinker -undefined -Xlinker dynamic_lookup -o ../lib.macosx-10.15-x86_64-3.8/modmesh/_modmesh.cpython-38-darwin.so CMakeFiles/_modmesh.dir/src/modmesh.cpp.o 
/Library/Developer/CommandLineTools/usr/bin/strip -x /Users/tigercosmos/modmesh/build/lib.macosx-10.15-x86_64-3.8/modmesh/_modmesh.cpython-38-darwin.so
[100%] Built target _modmesh
/usr/local/Cellar/cmake/3.18.4/bin/cmake -E cmake_progress_start /Users/tigercosmos/modmesh/build/temp.macosx-10.15-x86_64-3.8/CMakeFiles 0
```

**Comment by yungyuc** [comment]
Any difference in clang version?

**Review by yungyuc — APPROVED** [review]

**Comment by yungyuc** [comment]
It will also help to have cmake print compile command verbosely.

**Comment by tigercosmos** [comment]
I've checked. basically, LLVM, cmake, gcc are the same.

```
gcc path: /usr/bin/gcc
Configured with: --prefix=/Library/Developer/CommandLineTools/usr --with-gxx-include-dir=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/4.2.1
gcc version: Apple clang version 12.0.0 (clang-1200.0.32.21)
Target: x86_64-apple-darwin19.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
cmake path: /usr/local/bin/cmake
cmake version: cmake version 3.18.4

CMake suite maintained and supported by Kitware (kitware.com/cmake).
python3 path: /usr/local/bin/python3
python3 version: Python 3.8.5

numpy.__version__: 1.19.4

pytest path: /usr/local/bin/pytest

pytest 6.1.2
pytest version: 

clang-tidy path: /usr/local/bin/clang-tidy

clang-tidy version: LLVM (http://llvm.org/):
  LLVM version 11.0.0
  Optimized build.
  Default target: x86_64-apple-darwin19.6.0
  Host CPU: skylake

flake8 path: /usr/local/bin/flake8

flake8 version: 3.8.4 (mccabe: 0.6.1, pycodestyle: 2.6.0, pyflakes: 2.2.0) CPython 3.8.5 on
Darwin
```

**Comment by tigercosmos** [comment]
I think it's weird: it's not the business of clang-tidy to check libraries?

```
/Users/runner/work/modmesh/modmesh/include/modmesh/ConcreteBuffer.hpp:33:1: error: system include stdexcept not allowed, transitively included from /Users/runner/work/modmesh/modmesh/include/modmesh/ConcreteBuffer.hpp [llvmlibc-restrict-system-libc-headers,-warnings-as-errors]
#include <stdexcept>
^~~~~~~~~~~~~~~~~~~~
/Users/runner/work/modmesh/modmesh/include/modmesh/ConcreteBuffer.hpp:34:1: error: system include memory not allowed, transitively included from /Users/runner/work/modmesh/modmesh/include/modmesh/ConcreteBuffer.hpp [llvmlibc-restrict-system-libc-headers,-warnings-as-errors]
#include <memory>
^~~~~~~~~~~~~~~~~
/Users/runner/work/modmesh/modmesh/include/modmesh/ConcreteBuffer.hpp:54:16: error: 'make_shared<modmesh::ConcreteBuffer, unsigned long &, modmesh::ConcreteBuffer::ctor_passkey>' must resolve to a function declared within the '__llvm_libc' namespace [llvmlibc-callee-namespace,-warnings-as-errors]
        return std::make_shared<ConcreteBuffer>(nbytes, ctor_passkey());
               ^
/usr/local/Cellar/llvm/11.0.0/bin/../include/c++/v1/memory:4282:1: note: resolves to this declaration
make_shared(_Args&& ...__args)
^
/Users/runner/work/modmesh/modmesh/include/modmesh/ConcreteBuffer.hpp:57:65: error: 'construct' must resolve to a function declared within the '__llvm_libc' namespace [llvmlibc-callee-namespace,-warnings-as-errors]
    static std::shared_ptr<ConcreteBuffer> construct() { return construct(0); }
                                                                ^
/Users/runner/work/modmesh/modmesh/include/modmesh/ConcreteBuffer.hpp:33:1: error: system include stdexcept not allowed, transitively included from /Users/runner/work/modmesh/modmesh/include/modmesh/ConcreteBuffer.hpp [llvmlibc-restrict-system-libc-headers,-warnings-as-errors]
#include <stdexcept>
^~~~~~~~~~~~~~~~~~~~
/Users/runner/work/modmesh/modmesh/include/modmesh/ConcreteBuffer.hpp:34:1: error: system include memory not allowed, transitively included from /Users/runner/work/modmesh/modmesh/include/modmesh/ConcreteBuffer.hpp [llvmlibc-restrict-system-libc-headers,-warnings-as-errors]
#include <memory>
^~~~~~~~~~~~~~~~~
/Users/runner/work/modmesh/modmesh/include/modmesh/ConcreteBuffer.hpp:54:16: error: 'make_shared<modmesh::ConcreteBuffer, unsigned long &, modmesh::ConcreteBuffer::ctor_passkey>' must resolve to a function declared within the '__llvm_libc' namespace [llvmlibc-callee-namespace,-warnings-as-errors]
        return std::make_shared<ConcreteBuffer>(nbytes, ctor_passkey());
               ^
/usr/local/Cellar/llvm/11.0.0/bin/../include/c++/v1/memory:4282:1: note: resolves to this declaration
make_shared(_Args&& ...__args)
^
/Users/runner/work/modmesh/modmesh/include/modmesh/ConcreteBuffer.hpp:57:65: error: 'construct' must resolve to a function declared within the '__llvm_libc' namespace [llvmlibc-callee-namespace,-warnings-as-errors]
    static std::shared_ptr<ConcreteBuffer> construct() { return construct(0); }
                                                                  ^
```

**Comment by yungyuc** [comment]
Comparing the passing build: https://github.com/solvcon/modmesh/pull/11/checks?check_run_id=1395529801 and the failing one: https://github.com/solvcon/modmesh/pull/11/checks?check_run_id=1395529756 (against the same checkin), I believe `clang-tidy` is involved.

`clang-tidy` isn't a simple linter.

**Comment by tigercosmos** [comment]
how about on your mac?

**Comment by tigercosmos** [comment]
@yungyuc  The basic reason is that Ubuntu18 use llvm9, which is fine, but MacOS use llvm11, which break the CI.

**Comment by yungyuc** [comment]
Yep.

The last successful build ran with `clang-tidy` with LLVM 10: https://github.com/solvcon/modmesh/runs/1316331500?check_suite_focus=true#step:6:36

A day after, the Github Action mac runner switched to that with LLVM 11: https://github.com/solvcon/modmesh/runs/1322265793?check_suite_focus=true#step:6:36

This is the culprit.  Any thoughts about a fix?

File: .github/workflows/modmesh_install.yml [hunk 2]
```diff
@@ -13,7 +13,7 @@ jobs:
 
     strategy:
         matrix:
-          os: [ubuntu-18.04, macos-latest]
+          os: [ubuntu-20.04, macos-latest]
```
Comment (yungyuc) [inline .github/workflows/modmesh_install.yml]: I would like to keep 18.04.  It's OK to add 20.04, but please do not drop the older arch.

File: .github/workflows/modmesh_install.yml [hunk 3]
```diff
@@ -68,7 +68,7 @@ jobs:
     - name: build_ext
       run: |
         python3 setup.py build_ext \
-          --cmake-args="-DCMAKE_BUILD_TYPE=Release -DPYTHON_EXECUTABLE=$(which python3)" \
+          --cmake-args=" --tidy='/usr/local/bin/clang-tidy;-warnings-as-errors=*;--extra-arg-before=--driver-mode=g++' -DCMAKE_BUILD_TYPE=Release -DPYTHON_EXECUTABLE=$(which python3)" \
```
Comment (yungyuc) [inline .github/workflows/modmesh_install.yml]: Why don't you use `.clang-tidy` file instead?

**Review by yungyuc — CHANGES_REQUESTED** [review]

File: .github/workflows/modmesh_install.yml (hunk 3 above)
Comment (tigercosmos) [reply .github/workflows/modmesh_install.yml]: I am just trying. I will delete it later.

Comment (yungyuc) [reply .github/workflows/modmesh_install.yml]: It's better to test in your own branch without updating the PR.  It's confusing.

**Comment by tigercosmos** [comment]
@yungyuc  r?

**Comment by yungyuc** [comment]
@tigercosmos I don't understand what does "r?" mean.

**Review by yungyuc — APPROVED** [review]

File: .github/workflows/modmesh_pytest.yml [hunk 4]
```diff
@@ -41,9 +41,9 @@ jobs:
     - name: dependency (macos)
       if: matrix.os == 'macos-latest'
       run: |
-        brew install python3 llvm
-        ln -s "$(brew --prefix llvm)/bin/clang-format" "/usr/local/bin/clang-format"
-        ln -s "$(brew --prefix llvm)/bin/clang-tidy" "/usr/local/bin/clang-tidy"
+        brew install python3 llvm@9
```
Comment (yungyuc) [inline .github/workflows/modmesh_pytest.yml]: This doesn't seem to address the root cause.  The newer version `clang-tidy` (llvm 11) performs more checks and then `modmesh_pytest` fails.  Falling back to older version doesn't address the additional violation detected.

If the new violation is a true issue, it should get fixed.  If it's a false alarm or a fix is impractical, it should be explicitly excluded from the check.  Using older version is not the right approach to either of the treatments.

**Review by yungyuc — CHANGES_REQUESTED** [review]

File: .github/workflows/modmesh_pytest.yml (hunk 4 above)
Comment (tigercosmos) [reply .github/workflows/modmesh_pytest.yml]: r means review.
alright, then we should fix the code for llvm11.

Comment (yungyuc) [reply .github/workflows/modmesh_pytest.yml]: If it's easy to fix, we should fix it.  If you think it's a false alarm, let's discuss how to suppress it.

Comment (tigercosmos) [reply .github/workflows/modmesh_pytest.yml]: Yes, I think all of them are false alarms. I will postpone the fix to maybe next week since it takes me too much time today...
